### PR TITLE
Insertion Deletion AUC metric

### PR DESCRIPTION
**Pull request by GalyaZalesskaya**
- Added Insertion Deletion AUC metric that calculates the accuracy drop/increase during deletion/insertion of important pixels
image

- Added parent BasicMetric

### CODE REVIEW

File: openvino_xai/metrics/base.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+from abc import ABC, abstractmethod
+from typing import Any, Callable, List, Tuple
+
+import numpy as np
+import openvino.runtime as ov
```
Comment (goodsong81) [inline openvino_xai/metrics/base.py]:
```suggestion
import openvino as ov
```

File: openvino_xai/metrics/base.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+from openvino_xai.common.utils import IdentityPreprocessFN
+from openvino_xai.explainer.explanation import Explanation
+
+
+class BaseMetric(ABC):
+    """Base class for XAI quality metric."""
+
+    def __init__(
+        self,
+        model_compiled: ov.ie_api.CompiledModel = None,
```
Comment (goodsong81) [inline openvino_xai/metrics/base.py]:
```suggestion
        model_compiled: ov.CompiledModel = None,
```

File: openvino_xai/metrics/insertion_deletion_auc.py [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+import numpy as np
+
+from openvino_xai.explainer.explanation import Explanation, Layout
+from openvino_xai.metrics.base import BaseMetric
+
+
+def auc(arr):
+    """Returns normalized Area Under Curve of the array."""
+    return np.abs((arr.sum() - arr[0] / 2 - arr[-1] / 2) / (arr.shape[0] - 1))
```
Comment (goodsong81) [inline openvino_xai/metrics/insertion_deletion_auc.py]:
```suggestion
def AUC(arr):
    """Returns normalized Area Under Curve of the array."""
    return np.abs((arr.sum() - arr[0] / 2 - arr[-1] / 2) / (arr.shape[0] - 1))
```
Could you add the type annotation and doc string?

File: openvino_xai/metrics/base.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+    def model_predict(self, input: np.ndarray) -> np.ndarray:
+        logits = self.model_compiled([self.preprocess_fn(input)])
+        logits = self.postprocess_fn(logits)[0]
+        return logits
+
+    @abstractmethod
+    def evaluate(
+        self, explanations: List[Explanation], *args: Any, **kwargs: Any
+    ) -> float | Tuple[float, float, float]:
```
Comment (goodsong81) [inline openvino_xai/metrics/base.py]:
```suggestion
    ) -> Dict[str, float]:
```
What do you think about returning name, value pairs for generality?

**Review by goodsong81 — COMMENTED** [review]
Thank you for the update. I have a few comments below.
Besides, one alternative design idea came up in my mind. Could you please consider the following idea and my comments?

```python
class Metric(ABC):
    def __init__(
        self,
        score_fn: Callable[[np.ndarray], np.ndarray],  # image -> score vector
        # metric specific static parameters (like 'step' in AUC)
        *args,
        **kwargs,
    ):
        ...

    def __call__(  # Single image
        self,
        saliency_map: np.ndarray,
        class_idx: int,
        input_image: np.ndarray,
    ) -> Dict[str, float]:
        ...

    def evaluate(  # Dataset
        self,
        explanations: List[Explanation],
        input_images: List[np.ndarray],
    ) -> Dict[str, float]:
        ...
```

To me, setting model, preproc, postproc functions looks too complicated.
In above way, users could set arbitrary inference function that computes score vector from image. `Explainer.model_forward()` could also be reused in this way.

**Comment by GalyaZalesskaya** [comment]
Thank you for your suggestions, Songki. I implemented `__call__` for a single image and `evaluate` for a set of images. I also return `Dict[str, float]` for both of functions. It helps the generality but getting the score is no more elegant:

`pointing_game_score = self.pointing_game.evaluate([explanation], self.gt_bboxes)["pointing_game"]`
`pointing_game_score = list(self.pointing_game.evaluate([explanation], self.gt_bboxes).values())[0]`

Provably, `evaluate` can affect the internal state of `Metric` class and we can get scores as property function `pointing_game.score`. Let's leave the optimization until the next PR.

**Review by goodsong81 — APPROVED** [review]
Let's do the refactoring afterwards. Thank you for the update!
